WordPress/gutenberg
# Experimental page template selector

## Description

Introduces the foundations for Starter Page Templates (#18055). This PR introduces several pieces:

* A new `Page Templates` experiment to use as a feature flag for this feature while in development
* A `BlockListFooter` slot-fill to be able to conditionally insert the template picker at the end of the block list. This might change in the future as the native UI will likely be attached to the toolbar, and there's not enough clarity yet on what will happen on the web. For now, it offers a good starting point.
* A `withPageTemplatePickerVisible` HOC that encapsulates the logic on when the template picker should be visible (only on new/empty pages, if the experiment is enabled)
* A `PageTemplatePicker` component that shows the available templates (we hardcoded 2 for now) and replaces the post content when one is selected.

Fixes https://github.com/wordpress-mobile/gutenberg-mobile/issues/1464 and partially addresses https://github.com/wordpress-mobile/gutenberg-mobile/issues/1463 (the feature flag exists, but we'll need more changes in the bridge and initialization to make it work on mobile).

## How has this been tested?

On mobile:

- Use this PR: https://github.com/wordpress-mobile/gutenberg-mobile/pull/1576
- Launch the demo app. The template picker should be visible
- Tapping on a template should replace the page contents and make the picker disappear
- Tapping undo should remove the content and bring back the picker

On web:

- Go to Gutenberg > Experiments and enable the `Page Templates` experiment
- Create a new page. The template picker should be visible
- Pressing a template should replace the page contents and make the picker disappear
- Pressing undo should remove the content and bring back the picker
- Creating a new post should not show the template picker
- Opening a new post or page should not show the template picker





## Screenshots 

![spt-ios](https://user-images.githubusercontent.com/8739/68787082-316ba300-0641-11ea-857a-fdb4fbf0113c.png)

![Screen Shot 2019-11-13 at 18 24 02](https://user-images.githubusercontent.com/8739/68788064-f8343280-0642-11ea-85b0-d074c9eb7616.png)
![Screen Shot 2019-11-13 at 18 24 24](https://user-images.githubusercontent.com/8739/68788065-f8ccc900-0642-11ea-81e9-80023d92c985.png)
![Screen Shot 2019-11-13 at 18 24 26](https://user-images.githubusercontent.com/8739/68788066-f8ccc900-0642-11ea-8a30-2995e338775f.png)
![Screen Shot 2019-11-13 at 18 24 31](https://user-images.githubusercontent.com/8739/68788068-f8ccc900-0642-11ea-864e-b6643e008786.png)


## Types of changes

New feature

## Checklist:
